### PR TITLE
querier: Add a flag to limit time range for metadata APIs (Labels and Series)

### LINKED ISSUE #1811: querier: Limit LabelNames; LabelValues to certain time period.
Something that Cortex is doing is that `LabelNames;LabelValues` by default are limited to active series only on ingesters (so let's say in Thanos world to Thanos receiver or Prometheus for latest 2h). This is due to the fact that when you query you are usually interested in autocompletion for fresh metrics.

We are now exploring if we can apply the same logic for Thanos e.g limit `LabelNames` and `LabelValues` and `Series` (by default) on Querier QueryAPI to 2h. 

IF we don't do it, StoreGateway will have to load portions of the blocks in memory for full-time retention, for all store APIs, which is something we wanted to avoid. It was briefly mentioned here: https://github.com/thanos-io/thanos/issues/39 Also currently this is limited to raw data only, so it's already scoped for certain setups with retention for older blocks.

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

WIP

Ref #3119
Fixes #1811 

### Changes

* Add `query.metadata.default-time-range` flag to specify the default look-back duration for retrieving labels through Labels APIs when time range parameters are not specified.

### Verification

* `make test-local`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,9 +13,10 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 ### Added
 
-- [#3133](https://github.com/thanos-io/thanos/pull/3133) Query: Allow passing a `storeMatch[]` to Labels APIs. Also time range metadata based store filtering is supported on Labels APIs.
+- [#3133](https://github.com/thanos-io/thanos/pull/3133) Querier: Allow passing a `storeMatch[]` to Labels APIs. Also time range metadata based store filtering is supported on Labels APIs.
 - [#3154](https://github.com/thanos-io/thanos/pull/3154) Query Frontend: Add metric `thanos_memcached_getmulti_gate_queries_max`.
```
Comment (yeya24) [inline CHANGELOG.md:17]: Wait, why this is related to `Query Frontend` :laughing:  Would you mind updating it in this pr?

**Review by yeya24 — APPROVED** [review]
Looks perfect!

File: CHANGELOG.md (hunk 1 above)
Comment (kakkoyun) [reply CHANGELOG.md:17]: It doesn't change anything in `Query Frontend`. Just renamed a changelog entry.

Comment (yeya24) [reply CHANGELOG.md:17]: Why the changelog title is `Query Frontend`, I am confused... It is `pkg/gate`, so should be `Querier`?

Comment (kakkoyun) [reply CHANGELOG.md:17]: Now, I get it. I was focusing on the changes that I've made. I can fix it as well.

**Review by prmsrswt — APPROVED** [review]
LGTM!

**Comment by kakkoyun** [comment]
friendly ping @bwplotka

File: cmd/thanos/query.go [hunk 2]
```diff
@@ -81,6 +82,8 @@ func registerQuery(app *extkingpin.App) {
 
 	instantDefaultMaxSourceResolution := modelDuration(cmd.Flag("query.instant.default.max_source_resolution", "default value for max_source_resolution for instant queries. If not set, defaults to 0s only taking raw resolution into account. 1h can be a good value if you use instant queries over time ranges that incorporate times outside of your raw-retention.").Default("0s").Hidden())
 
+	defaultLabelLookbackDelta := cmd.Flag("query.labels.api-lookback-delta", "The default lookback duration for retrieving labels through Labels API when the range parameters are not specified. The zero value means range covers the time since the beginning.").Default("0s").Duration()
```
Comment (bwplotka) [inline cmd/thanos/query.go:85]: Ok, I would not overuse `lookback-delta` naming. This is really specific to PromQL and it means quite a specific thing -> not queried time range, but rather the time when the series is considered dead. Maybe just `query.metadata.default-time-range`?

File: pkg/api/query/v1.go [hunk 3]
```diff
@@ -659,3 +603,74 @@ func NewRulesHandler(client rules.UnaryClient, enablePartialResponse bool) func(
...
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0)
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999)
+)
+
+func parseLabelTimeRange(r *http.Request, defaultLabelLookbackDelta time.Duration) (time.Time, time.Time, error) {
+	// If start and end time not specified as query parameter, we get the range from the beginning of time by default.
+	var defaultStart, defaultEnd time.Time
+	if defaultLabelLookbackDelta == 0 {
+		defaultStart = minTime
```
Comment (bwplotka) [inline pkg/api/query/v1.go:616]: `infMinTime`? `as `minTime` is often used as `startTime`. Same for max (:

File: pkg/api/query/v1.go [hunk 4]
```diff
@@ -659,3 +603,74 @@ func NewRulesHandler(client rules.UnaryClient, enablePartialResponse bool) func(
...
+	start, err := parseTimeParam(r, "start", defaultStart)
+	if err != nil {
+		return time.Time{}, time.Time{}, &api.ApiError{Typ: api.ErrorBadData, Err: err}
+	}
+	end, err := parseTimeParam(r, "end", defaultEnd)
+	if err != nil {
+		return time.Time{}, time.Time{}, &api.ApiError{Typ: api.ErrorBadData, Err: err}
+	}
+	if end.Before(start) {
+		err := errors.New("end timestamp must not be before start time")
```
Comment (bwplotka) [inline pkg/api/query/v1.go:633]: https://thanos.io/tip/coding-style-guide.md/#avoid-defining-variables-used-only-once (:

File: cmd/thanos/query.go (hunk 2 above)
Comment (bwplotka) [reply cmd/thanos/query.go:85]: I propose `metadata` as I would love to cover `series` as well (on top of `labels, label_value`) I think it makes sense to group together, no need for flag for each of those :thinking: WDYT?

**Review by bwplotka — COMMENTED** [review]
Amazing, thanks! Some suggestion, otherwise LGTM :+1:

**Comment by kakkoyun** [comment]
@bwplotka Addressed the raised issues and included `series` API time range as well.

**Review by bwplotka — APPROVED** [review]
LGTM, :+1: 

Thanks! :muscle: 

I think it would be amazing to actually propose changing the default for this being ~1d or something in next release (:

**Review by yeya24 — APPROVED** [review]

**Pull request merged**